src/sender.rs

```diff
@@ -49,151 +53,218 @@ where
+        max_disk_bytes: usize,
+        mem_buffer: private::Queue<T>,
     ) -> Result<Sender<T>, super::Error>
     where
-        S: Into<String> + fmt::Display,
+        S: Into<String>,
     {
-        use std::sync::Arc;
-        let init_fs_lock = Arc::clone(&fs_lock);
-        let mut syn = init_fs_lock.lock().expect("Sender fs_lock poisoned");
+        let setup_mem_buffer = mem_buffer.clone(); // clone is cheeeeeap
+        let mut guard = setup_mem_buffer.lock_back();
         if !data_dir.is_dir() {
             return Err(super::Error::NoSuchDirectory);
         }
-        let seq_num = match fs::read_dir(data_dir)
-            .unwrap()
-            .map(|de| {
-                de.unwrap()
-                    .path()
-                    .file_name()
-                    .unwrap()
-                    .to_str()
-                    .unwrap()
-                    .parse::<usize>()
-                    .unwrap()
-            })
-            .max()
-        {
-            Some(sn) => sn,
-            None => 0,
-        };
-        let log = data_dir.join(format!("{}", seq_num));
-        match fs::OpenOptions::new().append(true).create(true).open(&log) {
-            Ok(fp) => {
-                syn.sender_fp = Some(BufWriter::new(fp));
-                (*syn).sender_seq_num = seq_num;
-                Ok(Sender {
-                    name: name.into(),
-                    root: data_dir.to_path_buf(),
-                    path: log,
-                    seq_num: seq_num,
-                    max_bytes: max_bytes,
-                    fs_lock: fs_lock,
-                    resource_type: PhantomData,
-                })
+        match private::read_seq_num(data_dir) {
+            Ok(seq_num) => {
+                let log = data_dir.join(format!("{}", seq_num));
+                match fs::OpenOptions::new().append(true).create(true).open(&log) {
+                    Ok(fp) => {
+                        (*guard).inner.sender_fp = Some(BufWriter::new(fp));
+                        (*guard).inner.sender_seq_num = seq_num;
+                        (*guard).inner.path = log;
+                        Ok(Sender {
+                            name: name.into(),
+                            root: data_dir.to_path_buf(),
+                            max_disk_bytes: max_disk_bytes,
+                            mem_buffer: mem_buffer,
+                            resource_type: PhantomData,
+                        })
+                    }
+                    Err(e) => Err(super::Error::IoError(e)),
+                }
             }
-            Err(e) => panic!("[Sender] failed to start {:?}", e),
+            Err(e) => Err(super::Error::IoError(e)),
         }
     }
 
+    fn write_to_disk(
+        &self,
+        event: T,
+        guard: &mut MutexGuard<BackGuardInner<SenderSync>>,
+    ) -> Result<(), (T, super::Error)> {
+        let mut buf: Vec<u8> = Vec::with_capacity(64);
+        buf.clear();
```

Comment:
This clear shouldn't be necessary, as the length of this vector is initially zero

Reply:
Ah, yep. Totally redundant.